src/box/lua/config/applier/credentials.lua

```diff
@@ -1,30 +1,326 @@
 local log = require('internal.config.utils.log')
 
-local function grant_privileges(name, privileges, role_or_user, grant_f)
-    for _, privilege in ipairs(privileges or {}) do
-        log.verbose('credentials.apply: grant %s to %s %s (if not exists)',
-            privilege, role_or_user, name)
-        for _, permission in ipairs(privilege.permissions or {}) do
-            local opts = {if_not_exists = true}
-            if privilege.universe then
-                grant_f(name, permission, 'universe', nil, opts)
-            end
-            -- TODO: It is not possible to grant a permission for
-            -- a non-existing object. It blocks ability to set it
-            -- from a config. Disabled for now.
-            --[[
-            for _, space in ipairs(privilege.spaces or {}) do
-                grant_f(name, permission, 'space', space, opts)
+--[[
+This intermediate representation is a formatted set of
+all permissions for a user or role. It is required to
+standardize diff functions. All the validation is done
+by config or box.info(), so neither the format nor the
+helper function don't perform it. Below you can find two
+converters to this representation, from box format and
+config schema format, accordingly `privileges_{box,config}_convert()`.
+
+[obj_type][obj_name] = {
+    read = true,
+    write = true,
+    ...
+}
+
+obj_types:
+ - 'user'
+ - 'role'
+ - 'space'
+ - 'function'
+ - 'sequence'
+ - 'universe'
+
+obj_names:
+ - mostly user defined strings, provided by config or box
+ - special value '', when there is no obj_name, e.g. for
+   'universe' obj_type or for granting permission for all
+   objects of a type.
```

Comment:
Raw idea (feel free to decline): we can use `universe` as the only available name of the object in the `universe` object type. It is similar how `nil` is the only object of the `nil` type.

The empty string fits here too.

Reply:
I don't want to have e.g. `'functions', 'all'` and `'universe'` being stored differently, so no.

Reply:
(The helper functions working with intermediate representation don't need to know the difference between object types (e.g. `'function'`, `'space'`, `'universe'`), when handling `'all'` case (`''` in ir))

Reply:
OK, got it.